src/Compilers/VisualBasic/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncMethodToClassRewriter.Spilling.vb

```diff
@@ -310,7 +315,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                             Return fieldAccess
                         End If
 
-                        Dim newReceiver As BoundExpression = SpillValue(fieldAccess.ReceiverOpt, isReceiver:=True, builder:=builder)
+                        ' An assignment target is only evaluated on write, so don't evaluate it's side effects, but do evaluate side effects of the reciever expression
+                        ' Evaluating a field of a struct has no side effects, so only evaluate side effects of the reciever expression
+                        Dim evaluateSideEffectsHere = evaluateSideEffects And Not isAssignmentTarget And fieldAccess.FieldSymbol.ContainingType.IsReferenceType
```

Comment:
>fieldAccess.FieldSymbol.ContainingType.IsReferenceType [](start = 105, length = 54)

Is this the right condition? Could this be a generic type parameter that is not known to be a reference type substituted with a reference type at runtime?

Reply:
It is not possible to assign to a field of a type parameter unless the type is constrained to a specific type. Only three type constraints have child types which can be reference or value types: `object`, `ValueType`, and `Enum`. None of them have fields.
However to be defensive we could check !IsValueType in case e.g. some runtime has a field on object, ValueType, or Enum. It will not cause any observable changes, but may lead to worse performance in such a case because we evaluate a side effect needlessly.

Reply:
> However to be defensive we could check !IsValueType in case e.g. some runtime has a field on object, ValueType, or Enum. It will not cause any observable changes, but may lead to worse performance in such a case because we evaluate a side effect needlessly

I don.t think there is a need for the change. I don't know what I was thinking. ```ContainingType``` is always a named type, never a type parameter. Therefore, we always know whether it is a reference type or not.

---
In reply to: [407885864](https://github.com/dotnet/roslyn/pull/42807#discussion_r407885864) [](ancestors = 407885864)